hive/src/test/java/org/apache/iceberg/hive/HiveCreateReplaceTableTest.java

```diff
@@ -0,0 +1,211 @@
+
+    Transaction txn = catalog.newCreateTableTransaction(
+        TABLE_IDENTIFIER, SCHEMA, SPEC, tableLocation, Maps.newHashMap());
+
+    // create the table concurrently
+    catalog.createTable(TABLE_IDENTIFIER, SCHEMA, SPEC);
+    Assert.assertTrue("Table should be created", catalog.tableExists(TABLE_IDENTIFIER));
+
+    // expect the transaction to fail
+    txn.commitTransaction();
+  }
+
+  @Test
+  public void testCreateTableTxnTableAlreadyExists() {
+    exceptionRule.expect(AlreadyExistsException.class);
+    exceptionRule.expectMessage("Table already exists: hivedb.tbl");
+
+    Assert.assertFalse("Table should not exist", catalog.tableExists(TABLE_IDENTIFIER));
+
+    // create a table before starting a transaction
+    catalog.createTable(TABLE_IDENTIFIER, SCHEMA, SPEC);
+    Assert.assertTrue("Table should be created", catalog.tableExists(TABLE_IDENTIFIER));
+
+    Transaction txn = catalog.newCreateTableTransaction(
+        TABLE_IDENTIFIER, SCHEMA, SPEC, tableLocation, Maps.newHashMap());
+    // expect the transaction to fail
+    txn.commitTransaction();
+  }
+
+  @Test
+  public void testReplaceTableTxn() {
+    catalog.createTable(TABLE_IDENTIFIER, SCHEMA, SPEC, tableLocation, Maps.newHashMap());
+    Assert.assertTrue("Table should exist", catalog.tableExists(TABLE_IDENTIFIER));
+
+    Transaction txn = catalog.newReplaceTableTransaction(
+        TABLE_IDENTIFIER, SCHEMA, PartitionSpec.unpartitioned(), tableLocation, Maps.newHashMap());
+    txn.commitTransaction();
+
+    Table table = catalog.loadTable(TABLE_IDENTIFIER);
+    Assert.assertEquals("Partition spec should match", PartitionSpec.unpartitioned(), table.spec());
+  }
+
+  @Test
+  public void testReplaceTableTxnTableNotExists() {
+    exceptionRule.expect(NoSuchTableException.class);
+    exceptionRule.expectMessage("No such table: hivedb.tbl");
+
+    Transaction txn = catalog.newReplaceTableTransaction(
+        TABLE_IDENTIFIER, SCHEMA, SPEC, tableLocation, Maps.newHashMap());
+    txn.updateProperties()
+        .set("prop", "value")
+        .commit();
+    // expect the transaction to fail
+    txn.commitTransaction();
+  }
+
+  @Test
+  public void testReplaceTableTxnTableDeletedConcurrently() {
+    exceptionRule.expect(NoSuchTableException.class);
+    exceptionRule.expectMessage("No such table: hivedb.tbl");
+
+    Assert.assertFalse("Table should not exist", catalog.tableExists(TABLE_IDENTIFIER));
+
+    catalog.createTable(TABLE_IDENTIFIER, SCHEMA, SPEC);
+    Assert.assertTrue("Table should be created", catalog.tableExists(TABLE_IDENTIFIER));
+
+    Transaction txn = catalog.newReplaceTableTransaction(
+        TABLE_IDENTIFIER, SCHEMA, PartitionSpec.unpartitioned(), tableLocation, Maps.newHashMap());
+    txn.updateProperties()
+        .set("prop", "value")
+        .commit();
+
+    // drop the table concurrently
+    catalog.dropTable(TABLE_IDENTIFIER);
+
+    // expect the transaction to fail
+    txn.commitTransaction();
```

Comment:
Here as well, I would expect this to succeed.

Maybe we should have a flag for createOrReplace?